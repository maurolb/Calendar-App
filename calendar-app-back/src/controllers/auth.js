const bcrypt = require('bcryptjs');
const jwtGenerator = require('../helpers/jwt');
const User = require('../models/User');

const createUser = async (userData) => {
  let user = await User.findOne({ email: userData.email });

  if (user) {
    return 'error email';
  }

  user = new User(userData);

  const salt = bcrypt.genSaltSync();
  user.password = bcrypt.hashSync(userData.password, salt);

  const { id, name } = await user.save();

  const token = await jwtGenerator(user.id, user.name);
  return { id, name, token };
};

const loginUser = async ({ email, password }) => {
  const user = await User.findOne({ email });

  if (!user) {
    return 'error email';
  }

  const validPassword = bcrypt.compareSync(password, user.password);

  if (!validPassword) {
    return 'error pass';
  }

  const token = await jwtGenerator(user.id, user.name);
  return { user, token };
};

const renewToken = async (uid, name) => {
  const newToken = await jwtGenerator(uid, name);
  return { newToken, uid, name };
};

module.exports = {
  createUser,
  loginUser,
  renewToken,
};
