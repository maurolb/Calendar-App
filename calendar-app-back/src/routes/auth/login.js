const { loginUser } = require('../../controllers/auth');

const login = async (req, res) => {
  try {
    const result = await loginUser(req.body);

    if (result === 'error email') {
      return res.status(400).json({ ok: false, message: 'Email incorrecto' });
    }
    if (result === 'error pass') {
      return res.status(400).json({ ok: false, message: 'Contrase√±a incorrecta' });
    }

    return res.status(200).json({
      ok: true, uid: result.user.id, name: result.user.name, token: result.token,
    });
  } catch (error) {
    console.log('error ', error.message);
    return res.status(500).json({ ok: false, message: 'Error interno, no se pudo logear el usuario' });
  }
};

module.exports = login;
