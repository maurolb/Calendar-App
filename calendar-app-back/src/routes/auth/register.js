const { createUser } = require('../../controllers/auth');

const register = async (req, res) => {
  try {
    const result = await createUser(req.body);
    if (result === 'error email') {
      return res.status(400).json({ ok: false, message: 'Ya existe un usuario con ese email' });
    }
    return res.status(201).json({
      ok: true, uid: result.id, name: result.name, token: result.token,
    });
  } catch (error) {
    console.log('error ', error.message);
    return res.status(500).json({ ok: false, message: 'Error interno, no se pudo crear el usuario' });
  }
};

module.exports = register;
