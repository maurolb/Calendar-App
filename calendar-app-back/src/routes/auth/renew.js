const { renewToken } = require('../../controllers/auth');

const renew = async (req, res) => {
  try {
    const result = await renewToken(req.uid, req.name);
    return res.status(200).json({
      ok: true, uid: result.uid, name: result.name, token: result.newToken,
    });
  } catch (error) {
    console.log('error ', error.message);
    return res.status(500).json({ ok: false, message: 'Error interno, no se pudo renovar el token' });
  }
};

module.exports = renew;
