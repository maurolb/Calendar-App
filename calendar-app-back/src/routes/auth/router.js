const { Router } = require('express');

const { check } = require('express-validator');
const renew = require('./renew');
const login = require('./login');
const register = require('./register');
const fieldValidators = require('../../middlewares/field-validators');
const jwtValidator = require('../../middlewares/jwt-validator');

const router = Router();

router.route('/')
  .post(
    [
      check('email', 'El email es requerido').isEmail(),
      check('password', 'La contraseña debe tener al menos 6 caracteres').isLength({ min: 6 }),
      fieldValidators,
    ],
    login,
  );

router.route('/register')
  .post(
    [
      check('name', 'El nombre es requerido').not().isEmpty(),
      check('email', 'El email es requerido').isEmail(),
      check('password', 'La contraseña debe tener al menos 6 caracteres').isLength({ min: 6 }),
      fieldValidators,
    ],
    register,
  );

router.route('/renew')
  .get(
    jwtValidator,
    renew,
  );

module.exports = router;
