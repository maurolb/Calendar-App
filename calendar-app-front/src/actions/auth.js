import Swal from "sweetalert2";
import { fetchConToken, fetchSinToken } from "../helpers/fetch";
import { types } from "../types/types";
import { eventLogout } from "./events";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const resp = await fetchSinToken('auth', {email, password}, 'POST');
    const body = await resp.json();
    
    if( body.ok ){
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime() );
      dispatch( login( {uid: body.uid, name: body.name} ))
    } else {
      Swal.fire('Error', body.message, 'error')
    }

  }
};

const login = ( user ) => ({
  type: types.authLogin,
  payload: user 
});

export const startRegister = (email, password, name) => {
  return async (dispatch) => {
    const resp = await fetchSinToken('auth/register', {email, password, name}, 'POST');
    const body = await resp.json();

    if( body.ok ){
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime() );
      dispatch( login( {uid: body.uid, name: body.name} ))
    } else {
      Swal.fire('Error', body.message, 'error')
    }

  }
};

export const startChecking = () => {
  return async (dispatch) => {
    const resp = await fetchConToken('auth/renew');
    const body = await resp.json();

    if( body.ok ){
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime() );

      dispatch( login( {uid: body.uid, name: body.name} ))

    } else {
      dispatch( checkingFinish() )
    }

  }
};

const checkingFinish = () => ({
  type: types.authFinishChecking
})

export const startLogout = () => {
  return async (dispatch) => {
    localStorage.clear();
    dispatch( eventLogout() );
    dispatch( logout() );
  }
}

const logout = () => ({
  type: types.authLogout
})