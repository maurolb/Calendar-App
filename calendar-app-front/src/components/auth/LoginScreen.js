import React from 'react';
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogin } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import './auth.css';

export const LoginScreen = () => {

  const dispatch = useDispatch();

  const [ formLoginValues, handleLoginInputChange ] = useForm({
    email: '',
    password: ''
  });

  const { email, password } = formLoginValues;

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch( startLogin(email, password) )
  }

  return (
    <div className="auth-container">
      <div className="login-form-1">
        <h3>Ingreso</h3>
        <form onSubmit={handleLogin}>
            <div className="form-group">
                <input 
                    type="text"
                    className="form-control"
                    placeholder="Correo"
                    name="email"
                    value={email}
                    onChange={handleLoginInputChange}
                />
            </div>
            <div className="form-group">
                <input
                    type="password"
                    className="form-control"
                    placeholder="ContraseÃ±a"
                    name="password"
                    value={password}
                    onChange={handleLoginInputChange}
                />
            </div>
            <div className="form-group">
                <input 
                    type="submit"
                    className="btnSubmit"
                    value="Login"
                />
            </div>
            <Link to="/auth/register" className="link">Registrate</Link>
          </form>
      </div>
    </div>
  )
}
