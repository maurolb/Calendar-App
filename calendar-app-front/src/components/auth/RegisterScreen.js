import React from 'react';
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import Swal from 'sweetalert2';
import { startRegister } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import './auth.css';

export const RegisterScreen = () => {

  const dispatch = useDispatch();

  const [ formRegisterValues, handleRegisterInputChange ] = useForm({
    name: '',
    email: '',
    password: '',
    password2: ''
  });

  const { name, email, password, password2 } = formRegisterValues;

  const handleRegister = (e) => {
    e.preventDefault();

    if( name === ''){
      return Swal.fire('Error', 'El nombre es requerido', 'error');
    }
    if( password !== password2 ){
      return Swal.fire('Error', 'Las contraseñas deben ser iguales', 'error');
    }
    dispatch( startRegister( email, password, name ) )
  }

  return (
    <div className='auth-container'>
      <div className="login-form-2">
        <h3>Registro</h3>
        <form onSubmit={handleRegister}>
          <div className="form-group">
              <input
                  type="text"
                  className="form-control"
                  placeholder="Nombre"
                  name="name"
                  value={name}
                  onChange={handleRegisterInputChange}
              />
          </div>
          <div className="form-group">
              <input
                  type="email"
                  className="form-control"
                  placeholder="Correo"
                  name="email"
                  value={email}
                  onChange={handleRegisterInputChange}
              />
          </div>
          <div className="form-group">
              <input
                  type="password"
                  className="form-control"
                  placeholder="Contraseña"
                  name="password"
                  value={password}
                  onChange={handleRegisterInputChange}
              />
          </div>

          <div className="form-group">
              <input
                  type="password"
                  className="form-control"
                  placeholder="Confirmar contraseña"
                  name="password2"
                  value={password2}
                  onChange={handleRegisterInputChange}
              />
          </div>

          <div className="form-group">
              <input 
                  type="submit" 
                  className="btnSubmit" 
                  value="Crear cuenta" />
          </div>
          <Link to="/auth/login" className="link">Ya estas registrado?</Link>
        </form>
      </div>
    </div>
  )
}
