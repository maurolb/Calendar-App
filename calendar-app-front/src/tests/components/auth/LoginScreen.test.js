import { mount } from "enzyme";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { startLogin } from "../../../actions/auth";
import { LoginScreen } from "../../../components/auth/LoginScreen";

jest.mock('../../../actions/auth', () => ({
  startLogin: jest.fn()
}));

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const initialState = {}

const store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <BrowserRouter>
      <LoginScreen />
    </BrowserRouter>
  </Provider>
)


describe('Tests on LoginScreen component', () => {
  
  beforeEach( () => {
    jest.clearAllMocks();
  });

  test('should show correctly', () => {
    expect( wrapper ).toMatchSnapshot();
  });

  test('should call login dispatch', () => {
    wrapper.find('input[name="email"]').simulate('change', {
      target: {
        name: 'email',
        value: 'test@testing.com'
      }
    });

    wrapper.find('input[name="password"]').simulate('change', {
      target: {
        name: 'password',
        value: 'asd123'
      }
    });

    // cuando existen 2 forms se selecciona el primero con .at(0)
    wrapper.find('form').at(0).simulate('submit', {
      preventDefault(){}
    });

    expect( startLogin ).toHaveBeenCalledWith("test@testing.com", "asd123")
  });
  

});
