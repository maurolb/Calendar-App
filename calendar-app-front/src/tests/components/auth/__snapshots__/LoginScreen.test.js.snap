// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on LoginScreen component should show correctly 1`] = `
<div
  className="auth-container"
>
  <div
    className="login-form-1"
  >
    <h3>
      Ingreso
    </h3>
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <input
          className="form-control"
          name="email"
          onChange={[Function]}
          placeholder="Correo"
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="form-control"
          name="password"
          onChange={[Function]}
          placeholder="ContraseÃ±a"
          type="password"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="btnSubmit"
          type="submit"
          value="Login"
        />
      </div>
      <a
        className="link"
        href="/auth/register"
        onClick={[Function]}
      >
        Registrate
      </a>
    </form>
  </div>
</div>
`;
